### DISCUSSION TITLE
Slice grid along polyline

### DISCUSSION THREAD

**Discussion by HansKallekleiv** [pyvista/pyvista#2397] [Q&A]
Hi,
I am trying to slice an `ExplicitStructuredGrid` using a polyline as input.
My use case is related to reservoir modelling, where the explicit structured grid represents a reservoir simulation model and the polyline represents a well. I want to make a vertical slice through the reservoir grid along the well trajectory and display with "unfolded" parallell projection

So far I have been playing with the `slice_along_line` function to create the slice, but seeing some unwanted behaviour.
- The slice is extended from first/last point in polyline to the edge of the grid
- The slice does not follow the polyline points exactly and some cells are left out.

As I'm new to both VTK and pyvista it is a bit unclear to me if this is expected behaviour, and if so if there is another way of doing this operation? 

See the picture/code below for a simple example.

Any input is appreciated :)

![image](https://user-images.githubusercontent.com/16436291/160803296-f3027cc1-51b2-4797-9a67-69de45ba3012.png)

```python
import pyvista as pv
import numpy as np

def create_explicit_structured_grid(ni: int, nj: int, nk: int, si: float, sj: float, sk: float) -> pv.ExplicitStructuredGrid:

    si = float(si)
    sj = float(sj)
    sk = float(sk)

    # create raw coordinate grid
    grid_ijk = np.mgrid[:(ni+1)*si:si, :(nj+1)*sj:sj, :(nk+1)*sk:sk]

    # repeat array along each Cartesian axis for connectivity
    for axis in range(1, 4):
        grid_ijk = grid_ijk.repeat(2, axis=axis)

    # slice off unnecessarily doubled edge coordinates
    grid_ijk = grid_ijk[:, 1:-1, 1:-1, 1:-1]

    # reorder and reshape to VTK order
    corners = grid_ijk.transpose().reshape(-1, 3)

    dims = np.array([ni, nj, nk]) + 1

    grid = pv.ExplicitStructuredGrid(dims, corners)
    grid = grid.compute_connectivity()
    grid.ComputeFacesConnectivityFlagsArray()

    return grid

grid = create_explicit_structured_grid(5, 4, 3, 20.0, 10.0, 5.0)


z = 14.0
line_verts = np.array(
    [
        (-5.0, -1.0, z),
        (5.0, 5.0, z),
        (6.0, 5.0, z),
        (7.0, 4.0, z),
        (8.0, 9.0, z),
        (15.0, 12.0, z),
        (25.0, 23.0, z),
        (70.0, 31.0, z),
        (65.0, 15.0, z),
        # (66.0, -2.0, z),
    ]
)

vertex_count = len(line_verts)
line_conn = np.arange(-1, vertex_count)
line_conn[0] = vertex_count

my_polyline = pv.PolyData(line_verts, lines=line_conn)

intersect = grid.slice_along_line(my_polyline)


plotter = pv.Plotter()
plotter.add_mesh(grid,style="wireframe",color="blue")
plotter.add_mesh(my_polyline,style="points", point_size=10, color="red")
plotter.add_mesh(my_polyline.copy(),style="surface", color="yellow")
plotter.add_mesh(intersect,color="blue")
plotter.show()
```

**Comment by MatthewFlamm** [comment]
This might be similar/related to https://github.com/pyvista/pyvista/issues/1305.  Your example might suggest that this slice filter only slices from cell boundary to cell boundary. That is it ignores all changes in the line inside a cell.  This makes sense, since we are 'slicing' the cell.  This would explain all the features you see in your example.  If so, this is a great reproducible example!

It seems like you might want to use the sampling filter.  The mesh you have is quite regular, so it should be straightforward to generate a sampling mesh.

**Reply by HansKallekleiv** [reply]
Thanks @MatthewFlamm !

Yes it looks very similar to that issue.

I will try your suggestion and report back.

The grid in this example was just to demonstrate though. The real grids are more irregular and can be quite large (millions of cells).
